File: plugins/welcome.lua
-- Welcome Plugin
-- Shows a welcome message and registers test commands

-- Show welcome message in status bar
editor.set_status("✨ Plugins are working! Welcome Plugin loaded successfully!")

-- Register commands that use built-in actions
editor.register_command({
    name = "Plugin Demo: Open Help",
    description = "Open the editor help page (uses built-in action)",
    action = "show_help",
    contexts = {"normal"}
})

editor.register_command({
    name = "Plugin Demo: Save File",
    description = "Save the current file (uses built-in action)",
    action = "save",
    contexts = {"normal"}
})

-- Register commands with custom Lua callbacks
editor.register_command({
    name = "Plugin Demo: Say Hello",
    description = "Insert a friendly greeting into the buffer",
    action = "plugin_say_hello",
    contexts = {"normal"},
    callback = function()
        editor.insert("👋 Hello from Lua! The plugin system is working!\n")
        editor.set_status("Inserted greeting at cursor position")
        debug("Plugin callback executed: say_hello")
    end
})

editor.register_command({
    name = "Plugin Demo: Insert Time",
    description = "Insert the current time at cursor position",
    action = "plugin_insert_time",
    contexts = {"normal"},
    callback = function()
        local time = os.date("%H:%M:%S")
        editor.insert("🕐 Current time: " .. time .. "\n")
        editor.set_status("Inserted time at cursor position")
        debug("Plugin callback executed: insert_time at " .. time)
    end
})

editor.register_command({
    name = "Plugin Demo: Insert Comment",
    description = "Insert a sample comment at cursor position",
    action = "plugin_insert_comment",
    contexts = {"normal"},
    callback = function()
        editor.insert("-- This comment was inserted by a Lua plugin!\n")
        editor.set_status("📝 Comment inserted by plugin")
        debug("Plugin callback executed: insert_comment")
    end
})

-- Debug output (goes to temp file, opens in background tab on first debug call)
debug("🎉 Welcome plugin initialized successfully!")
debug("📝 Registered 5 commands - try Ctrl+P to see them!")
debug("   - 'Plugin Demo: Open Help' - toggles help screen (built-in action)")
debug("   - 'Plugin Demo: Save File' - saves current file (built-in action)")
debug("   - 'Plugin Demo: Say Hello' - inserts greeting (Lua callback)")
debug("   - 'Plugin Demo: Insert Time' - inserts current time (Lua callback)")
debug("   - 'Plugin Demo: Insert Comment' - inserts sample comment (Lua callback)")
debug("")
debug("🔍 You're viewing this in the debug log! It opened automatically when the plugin called debug().")

File: plugins/welcome.ts
/// <reference path="./lib/fresh.d.ts" />
const editor = getEditor();


/**
 * Welcome Plugin
 * Shows a welcome message and registers demo commands
 */

// Show welcome message in status bar
editor.setStatus(editor.t("status.loaded"));

// Register commands that use built-in actions
editor.registerCommand(
  "%cmd.open_help",
  "%cmd.open_help_desc",
  "show_help",
  "normal"
);

editor.registerCommand(
  "%cmd.save_file",
  "%cmd.save_file_desc",
  "save",
  "normal"
);

// Register commands with custom TypeScript callbacks
globalThis.plugin_say_hello = function(): void {
  editor.insertAtCursor(editor.t("text.greeting") + "\n");
  editor.setStatus(editor.t("status.greeting_inserted"));
  editor.debug("Plugin callback executed: say_hello");
};

editor.registerCommand(
  "%cmd.say_hello",
  "%cmd.say_hello_desc",
  "plugin_say_hello",
  "normal"
);

globalThis.plugin_insert_time = function(): void {
  const time = new Date().toLocaleTimeString();
  editor.insertAtCursor(`Current time: ${time}\n`);
  editor.setStatus(editor.t("status.time_inserted"));
  editor.debug(`Plugin callback executed: insert_time at ${time}`);
};

editor.registerCommand(
  "%cmd.insert_time",
  "%cmd.insert_time_desc",
  "plugin_insert_time",
  "normal"
);

globalThis.plugin_insert_comment = function(): void {
  editor.insertAtCursor(editor.t("text.comment") + "\n");
  editor.setStatus(editor.t("status.comment_inserted"));
  editor.debug("Plugin callback executed: insert_comment");
};

editor.registerCommand(
  "%cmd.insert_comment",
  "%cmd.insert_comment_desc",
  "plugin_insert_comment",
  "normal"
);

// Debug output
editor.debug("Welcome plugin initialized successfully!");
editor.debug("Registered 5 commands - try Ctrl+P to see them!");
editor.debug("   - 'Plugin Demo: Open Help' - toggles help screen (built-in action)");
editor.debug("   - 'Plugin Demo: Save File' - saves current file (built-in action)");
editor.debug("   - 'Plugin Demo: Say Hello' - inserts greeting (TypeScript callback)");
editor.debug("   - 'Plugin Demo: Insert Time' - inserts current time (TypeScript callback)");
editor.debug("   - 'Plugin Demo: Insert Comment' - inserts sample comment (TypeScript callback)");
